docs/codegen-tour.md

```diff
@@ -0,0 +1,843 @@
+}
+```
+
+But this is not currently something we can express with Rust's generics.
+
+In any case: what this lets us do is avoid having two separate sets of types for
+the 'raw' and 'native' cases; we have a single wrapper type that we use anytime
+we want to indicate that a type is in its raw form. This has the additonal
+advantage that we can define new types in our generated code that implement
+`Scalar`, and then those types can automatically work with `BigEndian`; this is
+useful for things like custom enums and flags that are defined at various points
+in the spec.
+
+### `FixedSize`
+
+In addition to these two traits, we also have a [`FixedSize`][] trait, which is
+implemented for all scalar types (and later, for structs consisting only of
+scalar types). This trait consists of a single associated constant:
+
+```rust
+/// A trait for types that have a known, constant size.
+pub trait FixedSize: Sized {
+    /// The raw (encoded) size of this type, in bytes.
+    const RAW_BYTE_LEN: usize;
+}
+```
+
+This is implemented for both all the scalar values, as well as all their
+`BigEndian` equivalents; and in both cases, the value of `RAW_BYTE_LEN` is the
+size of the raw (big-endian) representation.
+
+### <a id="read-tables"></a> tables
+
+Conceptually, a table object is additional type information laid over a
+`FontData` object (a wrapper around a rust byte slice (`&[u8]`), essentially
+a pointer plus a length). It provides typed access to that tables fields.
+
+Conceptually, this looks like:
+
+```rust
+pub struct MyTable<'a>(FontData<'a>);
+
+impl MyTable<'_> {
+    /// Read the table's first field
+    pub fn format(&self) -> u16 {
+        self.0.read_at(0)
+    }
+}
+```
+
+In practice, what we generate is slightly different: instead of
+generating a struct for the table itself (and wrapping the data directly)
+we generate a 'marker' struct, which defines the type of the table, and then we
+combine it with the data via a `TableRef` struct.
+
+The `TableRef` struct looks like this:
+
+```rust
+/// Typed access to raw table data.
+pub struct TableRef<'a, T> {
+    shape: T,
+    data: FontData<'a>,
+}
+```
+
+And the definition of the table above, using a marker type, would look something
+like:
+
+```rust
+/// A marker type
+pub struct MyTableMarker;
+
+/// Instead of generating a struct for each table, we define a type alias
+pub type MyTable<'a> = TableRef<'a, MyTableMarker>;
+
+impl MyTableMarker {
+    fn format_byte_range(&self) -> Range<usize> {
+        0..u16::RAW_BYTE_LEN
+    }
+}
+
+impl MyTable<'_> {
+    fn format(&self) -> u16 {
+        let range = self.shape.format_byte_range();
+        self.data.read_at(range.start)
+    }
+}
+```
+
+To the user these two API are equivalent (you have a type `MyTable`, on which
+you can call methods to read fields) but the 'marker' pattern potentially allows
+for us to do some fancy things in the future (involving various cases where we
+want to store a type separate from a lifetime).
```

Comment:
I suppose this is the why, but ... it's pretty impenetrable for a Rust noob. I think we want it to make sense to a Rust noob.

Reply:
Unfortunately explaining the full rationale for this requires some familiarity with intermediate-to-advanced Rust. I could certainly go into more detail, but providing enough detail to fully explain the situation to a Rust noob would likely end up being another thousand words or so, and it ultimately is out of scope of the sort of high-level overview I'm trying to provide, here.

The idea I'm trying to express in this section is that it doesn't really matter *exactly* what types are involved in representing a table, right now; in either case the overall logic will be very similar.

Reply:
I would posit someone who bothers to read this IS interested in exactly what types and why. Maybe drop the makes sense to a Rust noob requirement if that lets it be both terse and informative. Right now it reads to me as hand wavily saying "there's some advanced stuff I can't be bothered to explain."